Some validators rework

lburg opened the pull request:
Taking feedback into account:

- drop `compound_validator` in favor of defining a list of validators to run;
- add a dedicated `'encoder'` field metadata attribute instead of mixing encoders with validators, as they're not the same.

lburg committed:
validators: Rename 'validator' field metadata attr 'validators'

Simple sed across the codebase to prepare a new feature: dropping
'compound_validator' in favor of specifying a list of validators
straight from the field definition.

lburg committed:
ocpp_codec: Replace compound_validator by defining several validators

Seems a bit easier to understand, and we could still give a name to
validators commonly used together, e.g.:

    is_strictly_positive = [is_positive, is_not_zero]

lburg committed:
serializer: Add a dedicated 'encoder' field to fields metadata

After some feedback, mixing encoders with validators turned out to be
confusing.

Encoders should now be assigned through the dedicated 'encoder' metadata
attribute. Encoders are defined in there own module, split tests
accordingly and sed code to match.

The notion of running validators+encoder is called cleaning, inspired
by Django. _run_validator is thus renamed _clean_data, and code and
comments are changed accordingly.

romainletendart commented on ocpp_codec/serializer.py:
```diff
@@ -34,20 +34,25 @@ def _required_fields(dataclass_class) -> typing.List[dataclasses.Field]:
 
 
 def _run_validator(field: dataclasses.Field, data: typing.Any, *, parsing: bool) -> typing.Any:
```
`_run_validators` with an `s` to follow new naming?

romainletendart commented on ocpp_codec/serializer.py:
```diff
@@ -34,20 +34,25 @@ def _required_fields(dataclass_class) -> typing.List[dataclasses.Field]:
+    validator_list = field.metadata.get('validators', [validators.noop])
+    # Defining a single validator is supported, turn it to a list for easier processing
+    if not isinstance(validator_list, list):
+        validator_list = [validator_list]
```
Having a metadata called `validators` in which we allow a non-list value sounds fishy to me.
I think we should have lists everywhere for consistency.

romainletendart replied:
Also, we could let users use any enumeration type like tuples which would me more appropriate in this case, what do you think?

rsebille replied:
Yeah, I agree on the list-everywhere thing.

romainletendart commented on ocpp_codec/v20/types.py:
```diff
@@ -255,27 +255,28 @@ class TriggerReasonEnum(utils.AutoNameEnum):
 class _IdentifierString20(types.SimpleType):
     value: str = field(metadata={
-        'validators': validators.compound_validator(validators.max_length_20, validators.is_identifier),
+        'validators': [validators.max_length_20, validators.is_identifier],
```
The new API is much nicer, well done!

rsebille commented on ocpp_codec/serializer.py:
```diff
@@ -34,20 +34,25 @@ def _required_fields(dataclass_class) -> typing.List[dataclasses.Field]:
-        # Pick the right method based on whether we're parsing a message, or serializing a field
-        validator = validator.from_json if parsing else validator.to_json
+    # Fetch the validators to run
+    validator_list = field.metadata.get('validators', [validators.noop])
```
```suggestion
    validator_list = field.metadata.get('validators', [])
```
I think an empty list works fine because it does the same thing than the following `validated_data = data` statement.

https://github.com/Polyconseil/ocpp-codec/blob/c5408459b374f133d2623f68ada75613e79b98ac/ocpp_codec/validators.py#L61-L63

rsebille commented on ocpp_codec/encoders.py:
```diff
@@ -0,0 +1,120 @@
+from ocpp_codec import errors
+
+
+class BaseEncoder:
```
:thinking: It may be one of the good reason to use a proper `abc.ABCMeta` and `@abstractmethod`.

lburg replied:
I've tried something, let me know what you think about it. It doesn't change much but it does make the intent clearer.

rsebille commented on ocpp_codec/serializer.py:
```diff
@@ -33,26 +34,30 @@ def _required_fields(dataclass_class) -> typing.List[dataclasses.Field]:
-    return validated_data
+
+    # Apply the defined encoder if it exists
+    encoder = field.metadata.get('encoder', encoders.NoopEncoder())
```
Why calling the `NoopEncoder` instead of returning early if there isn't one ?

lburg replied:
the idea was to keep a single point of exit in the function, but I think this might be so simple that I actually over complicate it with this class... I'll drop this, since I'm also dropping `valodators.noop` now that I just return an empty list of validators if none were defined.

rsebille reviewed (commented):
Thanks :heart:

lburg committed:
serializer: Enforce defining a list of validators, even for a single one

It's easier to understand since the key is named 'validators', plural,
seeing a list make sense, while you'd need to know an implicit
conversion is made when defining a single one.

Also, drop the noop validator since an empty list now does the same job.

This commit could have been merged with the previous ones, but I find it
clearer to act step by step here, since there's a lot of code moving
places, getting renamed, etc.

lburg committed:
encoders: Enforce overriding both base methods by marking them abstract

Simple API clarification, makes the intent extra clear.
